### PR TITLE
build: add more instructions to vendor management README

### PR DESCRIPTION
**Pull request by msbutler**
Previously, the instructions for installing and updating vendored dependencies
were hard to follow, especially for folks new to crbd. This commit adds
significantly more detail to the instructions on vendor dependency management.

Fixes: #71297

### CODE REVIEW

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71348)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msbutler)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMndTLAzCo4Vr4pia6:-MmMndTLAzCo4Vr4pia7:b-lq2upt) (no related file):*
Thanks for taking the time to make this more clear.  I've left some comments, but feel free to take or leave them.

---
___
*[build/README.md, line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-Mm3o7Lm6XJS_wqTnjsn:-Mm3o7Lm6XJS_wqTnjso:bqn5f57) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L158)):*
> ```Markdown
>    <dependency>@<version|branch|sha>`. You should see changes in `go.mod` when running `git diff`.
> 3. Add an import statement to a go file using `_ "[local file path to a 
>    package]"`. Adding this import statement ensures cockroach's make 
> ```

I'm a bit confused what you mean by `local file path` bit here.  You definitely need to add an import for the dependency for the subsequent instructions to work, but the only special thing about that import is that you might need to use an anonymous import if you haven't written any code referencing the dependency yet.
___
*[build/README.md, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMixsVDMGEaygTl7DF:-MmMixsVDMGEaygTl7DG:b-4pkvhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L200)):*
> ```
>     + Run `git add .`.  Note: run `git status` after you attempt to stage the changes-- the
>       vendored submodule sometimes fails to detect unstaged changes. To force its hand, run `git add -f "local_path_to_model"`, e.g.
>       "github.com/xitongsys/*"
> ```

Should `local_path_to_model` be `local_path_to_module`?

I imagine the "sometimes fails" bit here is perhaps because of gitignore or is there some other reason this might happen? If it is mostly `gitignore` issues, I wonder if "NOTE: Check `git status` after staging the changes as content may have been missed because of `.gitignore`. `git add -f PATH_TO_FILE` can be used to add git ignored files." would be clear enough.
___
*[build/README.md, line 210 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMhYLoEMkSEqaRDRTW:-MmMhYLoEMkSEqaRDRTX:b-qkb4v6) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L210)):*
> ```Markdown
> - Back in your `cockroachdb/cockroach` branch, commit your code changes and the new `vendor` 
>   submodule 
>   ref.
> ```

[nit] newlines/whitespace is a bit odd here.
___
*[build/README.md, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMjg3q8IzlaSGn9s1p:-MmMjg3q8IzlaSGn9s1q:bmqcwiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L217)):*
> ```Markdown
> 
>   + Organization members can push their named branches there directly, via `git push [remote 
>     vendor name, probs 'origin'] [your vendor branch] `
> ```

[nit] We might want to avoid breaking commands across lines in case people want to copy/paste them and aren't reading the rendered version. There is also an extra space at the end of the command here.
___
*[build/README.md, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMlQdU3Bvhppxm2B84:-MmMlQdV0ENBEiYtI47q:b-8fl04e) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L232)):*
I think this paragraph is less clear than the one it replaces.  How about something like:

> It is critical that any ref in `vendored` that is referenced from `cockroachdb/cockroach` remain available in `vendored` in perpetuity. One way to ensure this is to leave the `vendored` branch that you pushed your changes to in place.
> 
>  If you would like to delete your feature branch in the `vendored` repository, you must first ensure that another branch in `vendored` contains the commit referenced by `cockroachdb/cockroach`.  You can update the `master` branch in `vendored` to point at the git SHA currently referenced in `cockroachdb/cockroach`.

From what I've seen, we mostly just leave the feature branches in place, which I think is probably the easier default practice.
___
*[build/README.md, line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMf7nv8rN2c3zCK1ax:-MmMf7nv8rN2c3zCK1ay:bum7qrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L236)):*
I would maybe say:

> If you pull/rebase from `cockroach/cockroachdb` and encounter a conflict, it is often easiest to take the `master` branch's vendored directory and then recreate your vendor changes on top of it. For example:

For the directions below, I suppose we could alternatively put instructions on how to do the same thing but as part of the conflict resolution.  Note too that depending on what happened on `master` there may also be conflicts in go.mod and go.sum to resolve.  I typically take the same strategy there of preferring master's version and then rebuilding them post-rebase.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71348)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMndTLAzCo4Vr4pia6:-MmPE_KC01xz__Bx1VvJ:b8nz9t8) (no related file):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Thanks for taking the time to make this more clear.  I've left some comments, but feel free to take or leave them.
</blockquote></details>

@stevendanna thanks for all the comments! just pushed an update to the pr.

---
___
*[build/README.md, line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-Mm3o7Lm6XJS_wqTnjsn:-MmPEhRs57qGiPJL3aSZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L158)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I'm a bit confused what you mean by `local file path` bit here.  You definitely need to add an import for the dependency for the subsequent instructions to work, but the only special thing about that import is that you might need to use an anonymous import if you haven't written any code referencing the dependency yet.
</blockquote></details>

Done.
___
*[build/README.md, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMixsVDMGEaygTl7DF:-MmPEu3z2qGxQ3_VLi2R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L200)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

> ```
>     + Run `git add .`.  Note: run `git status` after you attempt to stage the changes-- the
>       vendored submodule sometimes fails to detect unstaged changes. To force its hand, run `git add -f "local_path_to_model"`, e.g.
>       "github.com/xitongsys/*"
> ```

Should `local_path_to_model` be `local_path_to_module`?

I imagine the "sometimes fails" bit here is perhaps because of gitignore or is there some other reason this might happen? If it is mostly `gitignore` issues, I wonder if "NOTE: Check `git status` after staging the changes as content may have been missed because of `.gitignore`. `git add -f PATH_TO_FILE` can be used to add git ignored files." would be clear enough.
</blockquote></details>

Done.
___
*[build/README.md, line 210 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMhYLoEMkSEqaRDRTW:-MmPEv-j4V5vqYRxx6qu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L210)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] newlines/whitespace is a bit odd here.
</blockquote></details>

Done.
___
*[build/README.md, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMjg3q8IzlaSGn9s1p:-MmPEvif8BTcxPbuosf3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L217)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] We might want to avoid breaking commands across lines in case people want to copy/paste them and aren't reading the rendered version. There is also an extra space at the end of the command here.
</blockquote></details>

Done.
___
*[build/README.md, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMlQdU3Bvhppxm2B84:-MmPF2J25epuTeY1vVjN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L232)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think this paragraph is less clear than the one it replaces.  How about something like:

> It is critical that any ref in `vendored` that is referenced from `cockroachdb/cockroach` remain available in `vendored` in perpetuity. One way to ensure this is to leave the `vendored` branch that you pushed your changes to in place.
> 
>  If you would like to delete your feature branch in the `vendored` repository, you must first ensure that another branch in `vendored` contains the commit referenced by `cockroachdb/cockroach`.  You can update the `master` branch in `vendored` to point at the git SHA currently referenced in `cockroachdb/cockroach`.

From what I've seen, we mostly just leave the feature branches in place, which I think is probably the easier default practice.
</blockquote></details>

Done.
___
*[build/README.md, line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71348#-MmMf7nv8rN2c3zCK1ax:-MmPF8QNEajjB9Ndlk6Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324c8831b8cbd568501b9b2fc5b52143c5a4f753/build/README.md#L236)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I would maybe say:

> If you pull/rebase from `cockroach/cockroachdb` and encounter a conflict, it is often easiest to take the `master` branch's vendored directory and then recreate your vendor changes on top of it. For example:

For the directions below, I suppose we could alternatively put instructions on how to do the same thing but as part of the conflict resolution.  Note too that depending on what happened on `master` there may also be conflicts in go.mod and go.sum to resolve.  I typically take the same strategy there of preferring master's version and then rebuilding them post-rebase.
</blockquote></details>

Done.

**Review by stevendanna — APPROVED** [review]
Thanks for updating this!  I left some non-blocking comments.

I think a challenge with this documentation is knowing what level of `git` and go modules knowledge to expect from the reader; let's see if this iteration helps the next person.

**Review by rickystewart — APPROVED** [review]
Thank you, this is a great improvement :)
